src/data/nat/digits.lean

```diff
@@ -0,0 +1,346 @@
+begin
+  induction L,
+  { refl, },
+  { dsimp [of_digits], rw L_ih, },
+end
+
+@[simp] lemma of_digits_one_cons {α : Type*} [semiring α] (h : ℕ) (L : list ℕ) :
+  of_digits (1 : α) (h :: L) = h + of_digits 1 L :=
+begin
+  dsimp [of_digits],
+  simp,
+end
+
+@[norm_cast] lemma coe_of_digits (α : Type*) [semiring α] (b : ℕ) (L : list ℕ) :
+  ((of_digits b L : ℕ) : α) = of_digits (b : α) L :=
+begin
+  induction L,
+  { refl, },
+  { dsimp [of_digits], push_cast, rw L_ih, }
+end
+
+@[norm_cast] lemma coe_int_of_digits (b : ℕ) (L : list ℕ) :
+  ((of_digits b L : ℕ) : ℤ) = of_digits (b : ℤ) L :=
+begin
+  induction L,
+  { refl, },
+  { dsimp [of_digits], push_cast, rw L_ih, simp, }
+end
+
+lemma digits_zero_of_eq_zero {b : ℕ} (h : 1 ≤ b) {L : list ℕ} (w : of_digits b L = 0) :
+  ∀ l ∈ L, l = 0 :=
+begin
+  induction L,
+  { intros l m,
+    cases m, },
+  { intros l m,
+    dsimp [of_digits] at w,
+    rcases m with rfl,
+    { convert nat.eq_zero_of_add_eq_zero_right w, simp, },
+    { exact L_ih ((nat.mul_right_inj h).mp (nat.eq_zero_of_add_eq_zero_left w)) _ m, }, }
+end
+
+lemma digits_of_digits
+  (b : ℕ) (h : 2 ≤ b) (L : list ℕ)
+  (w₁ : ∀ l ∈ L, l < b) (w₂ : if h : L = [] then true else L.last h ≠ 0):
+  digits b (of_digits b L) = L :=
+begin
+  induction L,
+  { dsimp [of_digits], simp },
+  { dsimp [of_digits],
+    rw dif_neg at w₂,
+    swap,
+    { simp, },
+    rw digits_add b h,
+    { rw L_ih,
+      { simp, },
+      { intros l m, apply w₁, exact list.mem_cons_of_mem _ m, },
+      { split_ifs,
+        { trivial, },
+        { rw [list.last_cons _ h_1] at w₂,
+            convert w₂, }}},
+    { convert w₁ L_hd (list.mem_cons_self _ _), simp, },
+    { by_cases h' : L_tl = [],
+      { rcases h' with rfl,
+        simp at w₂,
+        left,
+        apply nat.pos_of_ne_zero,
+        convert w₂, simp, },
+      { right,
+        apply nat.pos_of_ne_zero,
+        contrapose! w₂,
+        apply digits_zero_of_eq_zero _ w₂,
+        { rw list.last_cons _ h',
+          exact list.last_mem h', },
+        { exact le_of_lt h, }, }, }, },
+end
+
+lemma of_digits_digits (b n : ℕ) : of_digits b (digits b n) = n :=
+begin
+  cases b with b,
+  { cases n with n,
+    { refl, },
+    { change of_digits 0 [n+1] = n+1,
+      dsimp [of_digits],
+      simp, } },
+  cases b with b,
+  { induction n with n,
+    { refl, },
+    { simp [n_ih, add_comm 1], } },
+  apply nat.strong_induction_on n _, clear n,
+  intros n h,
+  cases n,
+  { refl, },
+  { change of_digits (b+2) ((n+1) % (b+2) :: digits (b+2) ((n+1) / (b+2))) = (n+1),
+    dsimp [of_digits],
+    rw h _ (nat.div_lt_self' n b),
+    simp [nat.mod_add_div], },
+end
+
+lemma of_digits_one (L : list ℕ) : of_digits 1 L = L.sum :=
+begin
+  induction L,
+  { refl, },
+  { dsimp [of_digits], simp [list.sum_cons, L_ih], }
+end
+
+-- I don't want to move this to `euclidean_domain.lean`, because `ring` isn't available there.
+lemma dvd_mul_sub_mul {α : Type*} [euclidean_domain α]
```

Comment:
Why would you want to assume `euclidean_domain` instead of just `comm_ring`? Also, with the proof
```lean
begin
  convert dvd_add (dvd_mul_of_dvd_right hxy a) (dvd_mul_of_dvd_left hab y),
  rw [mul_sub_left_distrib, mul_sub_right_distrib],
  simp only [sub_eq_add_neg, add_assoc, neg_add_cancel_left],
end
```
you can put it where it should be.

Reply:
Fixed, thanks!